### PR TITLE
Tc 45 jest tests part 2

### PR DESCRIPTION
**Pull request by rizwan171**
Another batch of jest tests. I've been trying to PR them in small batches but this one is still a bit big so apologies.

To test:
Run `yarn test` and confirm all tests pass

### CODE REVIEW

File: src/tests/utils/stateUtils.tsx [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import { ModalActionMenuVisibilityState } from "../../types/reducers/ModalActionMenuVisibilitySlice";
+
+export function generateModalActionsVisibilityState(): ModalActionMenuVisibilityState {
```
Comment (rizwan171) [inline src/tests/utils/stateUtils.tsx:3]: As part of the tests for components under CardModalActions, you need to specify he ModalActionMenuVisibility state object. In each test, we only check that one menu is open, and it's pretty tedious and confusing to have to define the whole state in the test. Therefore, I created this helper to generate the default state, and any modifications to this state can happen in the test, reducing the amount of code we write and the amount of unnecessary code in the tests

File: docs/Testing Information.md [hunk 2]
```diff
@@ -302,11 +322,44 @@ it("should do something", () => {
   fireEvent.click(button);
 
   // will not be true
-  waitFor(() => {
+  await waitFor(() => {
```
Comment (rizwan171) [inline docs/Testing Information.md:325]: waitFor is async, so must be awaited

File: src/components/CardModal/CardModalActions/AttachmentMenu/AttachmentMenu.tsx [hunk 3]
```diff
@@ -21,13 +21,15 @@ const AttachmentMenu = ({ update }: AttachmentMenuProps) => {
...
+    const setFilePromises = [];
     if (files) {
       for (let i = 0; i < files.length; i++) {
         const id = uuidv4();
         const date = JSON.stringify(new Date());
-        await localforage.setItem(id, files[i]);
+        setFilePromises.push(localforage.setItem(id, files[i]));
         upload.push({ id, date, name: files[i].name });
       }
+      await Promise.all(setFilePromises);
```
Comment (rizwan171) [inline src/components/CardModal/CardModalActions/AttachmentMenu/AttachmentMenu.tsx:32]: I've changed this slightly to make testing easier. Granted, this isn't a best practice but when testing this, it's hard to check the date is what we expect as the await inside the loop causes the date to be different than the mock. I couldn't find a solution in jest for this (though I'm sure there is one), so in the interest of saving time and progressing, I've changed this logic slightly to gather all the setItem promises and await all of them once the for loop is done.

File: src/components/CardModal/CardModalActions/AttachmentMenu/AttachmentMenu.tsx [hunk 4]
```diff
@@ -42,8 +44,13 @@ const AttachmentMenu = ({ update }: AttachmentMenuProps) => {
...
       <div className="relative text-center mb-2">
-        <span className="text-sm block relative z-10">Attach From... </span>
-        <MdClose onClick={handleClose} size={20} className="absolute right-0 top-0 z-20 cursor-pointer" />
+        <span className="text-sm block relative z-10">Attach From...</span>
+        <MdClose
+          data-testid="attachment-menu-close"
+          onClick={handleClose}
+          size={20}
+          className="absolute right-0 top-0 z-20 cursor-pointer"
+        />
```
Comment (rizwan171) [inline src/components/CardModal/CardModalActions/AttachmentMenu/AttachmentMenu.tsx:53]: Formatting change here as well as a `data-testid` attribute

File: src/components/CardModal/CardModalActions/AttachmentMenu/AttachmentMenu.tsx [hunk 5]
```diff
@@ -53,7 +60,15 @@ const AttachmentMenu = ({ update }: AttachmentMenuProps) => {
...
-        <input multiple className="hidden" type="file" name="file" onChange={handleFileSelect} ref={hiddenFileInput} />
+        <input
+          data-testid="attachment-menu-file-upload"
+          multiple
+          className="hidden"
+          type="file"
+          name="file"
+          onChange={handleFileSelect}
+          ref={hiddenFileInput}
+        />
```
Comment (rizwan171) [inline src/components/CardModal/CardModalActions/AttachmentMenu/AttachmentMenu.tsx:71]: same as above

File: src/components/CardModal/CardModalActions/CopyMenu/CopyMenu.tsx [hunk 6]
```diff
@@ -25,7 +25,12 @@ const CopyMenu = () => {
     <div className="fixed mt-10 w-72 text-gray-700 bg-white rounded-ibsm shadow-2xl p-4">
       <div className="relative text-center mb-2">
         <span className="text-sm block relative z-10">Copy Card</span>
-        <MdClose onClick={handleClose} size={20} className="absolute right-0 top-0 z-20 cursor-pointer" />
+        <MdClose
+          data-testid="close-icon"
+          onClick={handleClose}
+          size={20}
+          className="absolute right-0 top-0 z-20 cursor-pointer"
+        />
```
Comment (rizwan171) [inline src/components/CardModal/CardModalActions/CopyMenu/CopyMenu.tsx:33]: Formatting change here as well as a data-testid attribute

File: src/components/CardModal/CardModalActions/MoveMenu/MoveMenu.tsx [hunk 7]
```diff
@@ -19,14 +20,20 @@ const MoveMenu = () => {
     if (!card) return;
 
     dispatch(moveCardToList({ cardId: card.id, destListId: selectedListId }));
+    dispatch(setCurrentSelectedCard({ ...card, listId: selectedListId }));
```
Comment (rizwan171) [inline src/components/CardModal/CardModalActions/MoveMenu/MoveMenu.tsx:23]: This is quite important and something I missed. `currentSelectedCard` controls the card data associated with the card modal. If this is out of sync at any point, updates may not be reflected correctly. 

Previously it updated the cards state to update the list id of the card. However, this doesn't automatically update `currentSelectedCard`. To do that, you need to set it

File: src/components/CardModal/CardModalActions/MoveMenu/MoveMenu.tsx [hunk 8]
```diff
@@ -19,14 +20,20 @@ const MoveMenu = () => {
...
     <div className="fixed mt-10 w-72 text-gray-700 bg-white rounded-ibsm shadow-2xl p-4">
       <div className="relative text-center mb-2">
         <span className="text-sm block relative z-10">Move Card</span>
-        <MdClose onClick={handleClose} size={20} className="absolute right-0 top-0 z-20 cursor-pointer" />
+        <MdClose
+          data-testid="close-icon"
+          onClick={handleClose}
+          size={20}
+          className="absolute right-0 top-0 z-20 cursor-pointer"
+        />
```
Comment (rizwan171) [inline src/components/CardModal/CardModalActions/MoveMenu/MoveMenu.tsx:36]: Formatting change here as well as a data-testid attribute

File: src/tests/components/BoardOptionsMenu/ExportModal/ExportModal.test.tsx [hunk 9]
```diff
@@ -45,13 +45,11 @@ describe("ExportModal", () => {
...
 
-    const expectedOptionValues = initialState.lists?.value.map((option) => option.id);
-    expectedOptionValues?.unshift("");
+    const expectedOptionValues = ["", "1", "2", "3", "4", "5"];
     const actualOptionValues = [...selector.options].map((option) => option.value);
     expect(actualOptionValues).toStrictEqual(expectedOptionValues);
 
-    const expectedOptionTexts = initialState.lists?.value.map((option) => option.title);
-    expectedOptionTexts?.unshift("Select List...");
+    const expectedOptionTexts = ["Select List...", "Test 1", "Test 2", "Test 3", "Test 4", "Test 5"];
```
Comment (rizwan171) [inline src/tests/components/BoardOptionsMenu/ExportModal/ExportModal.test.tsx:52]: I decided to change this from being calculated from variables to explicit strings. It's much clearer to see what we expect at a quick glance in my opinion

File: src/tests/components/CardModal/CardModalActions/AttachmentMenu/AttachmentMenu.test.tsx [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+    const stateBeforeUpload = store.getState();
+    expect(stateBeforeUpload.currentSelectedCard.value?.attachments).toHaveLength(0);
+    expect(stateBeforeUpload.modalActionMenusVisibility.value.attachmentMenuOpen.status).toBeTruthy();
+
+    const hiddenFileInput = screen.getByTestId("attachment-menu-file-upload") as HTMLInputElement;
+    expect(hiddenFileInput).toBeInTheDocument();
+
+    const uploadedFile = new File(["This is a test file"], "test.txt", { type: "text/plain" });
+    const date = new Date();
+    jest.useFakeTimers().setSystemTime(date);
```
Comment (rizwan171) [inline src/tests/components/CardModal/CardModalActions/AttachmentMenu/AttachmentMenu.test.tsx:104]: This fake timers is needed here as without it, we cannot check the date that is set on the attachment. Using jest to fake the current time ensures that any calls to new Date() will return the `date` variable, meaning we can check the attachment is created with a date correctly and it matches what we expect

**Review by nk711 — APPROVED** [review]
Looks good to me. The tests are well structured and were easy to read.

**Pull request merged**